File: section3/section 3.1/stamps.py
"""
ID: shikha11
LANG: PYTHON3
TASK: stamps
"""

fin = open('stamps.in', 'r')
fout = open('stamps.out', 'w')

K, N = map(int, fin.readline().strip().split())

while True:
    line = fin.readline()
    if not line or not line[0].isdigit():
        break
    stamps = list(map(int, line.strip().split()))

max_stamp = max(stamps)
max_value = max_stamp * K

# Initialize a 1D array to store the minimum number of coins needed for each value
dp = [float('inf')] * (max_value + 1)
dp[0] = 0

for coin in stamps:
    for value in range(coin, max_value + 1):
        dp[value] = min(dp[value], dp[value - coin] + 1)

n = 0
while n <= max_value and dp[n] <= K:
    n += 1

fout.write(str(n - 1) + '\n')

# max_stamp = max(stamps)
# max_value = max_stamp * K
#
#
# possible = [[False] * (max_value + 1) for _ in range(2)]
# possible[0][0] = True
# for stamp in stamps:
#     possible[0][stamp] = True
#
# c = 1
# for num_coins in range(2, K + 1):
#     for value in range(0, num_coins * max_stamp + 1):
#         possible[c][value] = possible[c - 1][value]
#
#         # if it's possible with lesser number of coins, then move on
#         if possible[c][value]:
#             continue
#
#         # otherwise check if adding any coin will make it possible
#         for coin in stamps:
#             if value - coin >= 0 and possible[c - 1][value - coin]:
#                 possible[c][value] = True
#                 break
#
#     possible[c-1] = possible[1]
#     possible[1] = [False] * (max_value + 1)
#
# n = 0
# while n <= max_value:
#     if possible[0][n]:
#         n += 1
#     else:
#         break
#
# fout.write(str(n - 1) + '\n')

File: section3/section 3.1/stamps.java
/*
ID: shikha11
LANG: JAVA
TASK: stamps
 */

import java.io.BufferedReader;
import java.io.FileReader;
import java.io.FileWriter;
import java.io.IOException;

public class stamps {
  public static void main(String[] args) throws IOException {
    BufferedReader fin = new BufferedReader(new FileReader("stamps.in"));
    FileWriter fout = new FileWriter("stamps.out");

    String[] kn = fin.readLine().split(" ");
    int K = Integer.parseInt(kn[0]);
    int N = Integer.parseInt(kn[1]);

    int maxStamp = 0;
    int[] stamps = new int[N];
    String line;
    while ((line = fin.readLine()) != null && Character.isDigit(line.charAt(0)) && maxStamp < N) {
      String[] tokens = line.split(" ");
      for (String token : tokens) {
        int intValue = Integer.parseInt(token);
        stamps[maxStamp] = intValue;
        maxStamp++;
      }
    }

    int maxStampValue = 0;
    for (int i = 0; i < maxStamp; i++) {
      maxStampValue = Math.max(maxStampValue, stamps[i]);
    }
    int maxPossibleValue = maxStampValue * K;

    // initialize dp array to infinity
    int[] dp = new int[maxPossibleValue + 1];
    for (int i = 1; i <= maxPossibleValue; i++) {
      dp[i] = 200000000;
    }

    for (int i = 0; i < maxStamp; i++) {
      for (int j = stamps[i]; j <= maxPossibleValue; j++) {
        dp[j] = Math.min(dp[j], dp[j - stamps[i]] + 1);
      }
    }

    int result = 0;
    while (result <= maxPossibleValue && dp[result] <= K) {
      result++;
    }

    fout.write(String.valueOf(result - 1) + "\n");
    fout.close();
  }
}
